br/pkg/lightning/backend/external/byte_reader.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"context"
+	"io"
+
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/mathutil"
+	"go.uber.org/zap"
+)
+
+// byteReader provides structured reading on a byte stream of external storage.
+type byteReader struct {
+	ctx           context.Context
+	storageReader storage.ReadSeekCloser
+
+	buf       []byte
+	bufOffset int
+
+	auxBuf []byte
+	isEOF  bool
+
+	retPointers []*[]byte
+}
+
+func openStoreReaderAndSeek(
+	ctx context.Context,
+	store storage.ExternalStorage,
+	name string,
+	initFileOffset uint64,
+) (storage.ReadSeekCloser, error) {
+	storageReader, err := store.Open(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+	_, err = storageReader.Seek(int64(initFileOffset), io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+	return storageReader, nil
+}
+
+func newByteReader(ctx context.Context, storageReader storage.ReadSeekCloser, bufSize int) *byteReader {
+	return &byteReader{
+		ctx:           ctx,
+		storageReader: storageReader,
+		buf:           make([]byte, bufSize),
+		bufOffset:     0,
+	}
+}
+
+// readNBytes reads the next n bytes from the reader and returns a buffer slice containing those bytes.
+// The returned slice (pointer) can not be used after r.reset. In the same interval of r.reset,
+// byteReader guarantees that the returned slice (pointer) will point to the same content
+// though the slice may be changed.
+func (r *byteReader) readNBytes(n int) (*[]byte, error) {
+	b := r.next(n)
+	readLen := len(b)
+	if readLen == n {
+		ret := &b
+		r.retPointers = append(r.retPointers, ret)
+		return ret, nil
+	}
+	// If the reader has fewer than n bytes remaining in current buffer,
+	// `auxBuf` is used as a container instead.
+	if cap(r.auxBuf) < n {
+		r.auxBuf = make([]byte, n)
+	}
+	r.auxBuf = r.auxBuf[:n]
+	copy(r.auxBuf, b)
+	for readLen < n {
+		r.cloneSlices()
+		err := r.reload()
+		if err != nil {
+			return nil, err
+		}
+		b = r.next(n - readLen)
+		copy(r.auxBuf[readLen:], b)
+		readLen += len(b)
+	}
+	return &r.auxBuf, nil
```

Comment:
the returned auxBuf is not cloned so its content will be changed. I'll check it by a test soon

Reply:
```golang
func TestAuxBuf(t *testing.T) {
	ms := &mockExtStore{src: []byte("0123456789")}
	br := newByteReader(context.Background(), ms, 1)
	require.NoError(t, br.reload())
	y1, err := br.readNBytes(1)
	require.NoError(t, err)
	y2, err := br.readNBytes(2)
	require.NoError(t, err)
	require.Equal(t, []byte("0"), *y1)
	require.Equal(t, []byte("12"), *y2)

	y3, err := br.readNBytes(1)
	require.NoError(t, err)
	y4, err := br.readNBytes(2)
	require.NoError(t, err)
	require.Equal(t, []byte("3"), *y3)
	require.Equal(t, []byte("45"), *y4)
	require.Equal(t, []byte("0"), *y1)
	require.Equal(t, []byte("12"), *y2)
}
```

Seems an irrelevant fail